### PR TITLE
Refactor/addresses sdk

### PR DESCRIPTION
**Pull request by Kiryous**

### CODE REVIEW

File: packages/dialect-react-sdk/hooks/useAddresses.ts [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+      if (!isWalletConnected || !dappPublicKey || !address) return;
+      setVerifyingCode(true);
+      try {
+        const data = await walletsApi.addresses.verify({
+          addressId: address.id,
+          code,
+        });
+        await mutateAddresses(mergeAddress(data));
+      } catch (err) {
+        throw err as Error;
```
Comment (fsher) [inline packages/dialect-react-sdk/hooks/useAddresses.ts:334]: I see a lot of similar casts to `Error`. May I ask why is that necessary?

File: packages/dialect-react-ui/components/Notifications/index.tsx [hunk 2]
```diff
@@ -35,111 +37,153 @@ interface NotificationsProps {
...
 
-  const { navigate } = useRoute();
+      if (
+        !wallet?.publicKey ||
+        isFetchingAddresses ||
+        isFetchingThread ||
+        isCreatingThread ||
+        isDeletingThread
+      )
+        return;
```
Comment (fsher) [inline packages/dialect-react-ui/components/Notifications/index.tsx:98]: could you please add a block statement for this `return`. Otherwise it looks like it's floating :D

File: packages/dialect-react-ui/components/Notifications/screens/Settings/CreateNotificationsThread.tsx [hunk 3]
```diff
@@ -58,33 +52,45 @@ const CreateNotificationsThread = ({
...
+        encrypted: false,
+        backend,
       });
-  }, [backend, create, dappAddress, onThreadCreated, onThreadCreationFailed]);
+
+      // Address would be created by syncState effect
+
+      await showThread(thread);
+    } catch (e) {
+      // TODO: do we need to do smth here?
```
Comment (fsher) [inline packages/dialect-react-ui/components/Notifications/screens/Settings/CreateNotificationsThread.tsx:91]: I'd check if `onThreadCreationFailed` was used somewhere or somehow.

**Review by fsher — APPROVED** [review]
Looking good, sir! Approving in advance, left minor comments

File: packages/dialect-react-sdk/hooks/useAddresses.ts [hunk 4]
```diff
@@ -0,0 +1,398 @@
...
+            addressIds: addresses?.map((addr) => addr.id),
+          })
+      : null,
+    {
+      refreshInterval,
+      refreshWhenOffline: true,
+    }
+  );
+
+  const isFetchingDappAddresses =
```
Comment (zhelezkov) [inline packages/dialect-react-sdk/hooks/useAddresses.ts:122]: can you explain why is this like this?

In my mind this doesn't reflect what really happens
I understand the case you want to handle on the ui side, but in my opinion this should reflect the real process behind api

File: packages/dialect-react-sdk/hooks/useAddresses.ts [hunk 5]
```diff
@@ -0,0 +1,398 @@
...
+  | ToggleParams
+  | VerifyParams
+  | ResendParams;
+
+type AddressEnriched = Address & {
+  enabled: boolean;
+  dappAddress?: DappAddress;
+};
+interface UseAddressesValue {
+  addresses: Record<AddressType, AddressEnriched> | Record<string, never>;
```
Comment (zhelezkov) [inline packages/dialect-react-sdk/hooks/useAddresses.ts:52]: why type like this?

File: packages/dialect-react-sdk/hooks/useThread.ts [hunk 6]
```diff
@@ -78,11 +77,11 @@ const useThread = ({
 
   return {
     // sdk
-    thread,
+    thread: thread || null,
     delete: deleteThread,
 
     // react-lib
-    isFetchingThread,
+    isFetchingThread: thread === undefined && !errorFetchingThread,
```
Comment (zhelezkov) [inline packages/dialect-react-sdk/hooks/useThread.ts:84]: see comment above, wondering why it should be like this if it doesn't reflect what's really happens in the api

File: packages/dialect-react-ui/components/Notifications/index.tsx (hunk 2 above)
Comment (zhelezkov) [reply packages/dialect-react-ui/components/Notifications/index.tsx:98]: think we should ban this style in eslint haha

**Review by zhelezkov — CHANGES_REQUESTED** [review]
this is good, great work. left some open questions to discuss

File: packages/dialect-react-sdk/hooks/useThread.ts (hunk 6 above)
Comment (Kiryous) [reply packages/dialect-react-sdk/hooks/useThread.ts:84]: It actually reflects the loading state what happens in API, better then isValidating. See SWR docs, 

https://swr.vercel.app/docs/getting-started#make-it-reusable.

The isFetching would be true every time the request is made, it will result into flickering loading state every pollingInterval. 

What we need instead just the initial loading.

File: packages/dialect-react-sdk/hooks/useAddresses.ts (hunk 5 above)
Comment (zhelezkov) [reply packages/dialect-react-sdk/hooks/useAddresses.ts:52]: My main concern is about `Record<string, never>`

File: packages/dialect-react-sdk/hooks/useAddresses.ts [hunk 7]
```diff
@@ -0,0 +1,398 @@
...
+  refreshInterval?: number;
+}
+
+function useAddresses({
+  refreshInterval,
+}: UseAddressesParams = EMPTY_OBJ): UseAddressesValue {
+  const { wallet: walletsApi } = useDialectSdk();
+  const { dappAddress: dappPublicKey } = useDialectDapp();
+  const { connected: isWalletConnected } = useDialectWallet();
+  const [isCreatingAddress, setCreatingAddress] = useState(false);
```
[the remark this answers is not in the document]
Comment (Kiryous) [reply packages/dialect-react-sdk/hooks/useAddresses.ts:81]: hey! thanks for the proposal, I don't want to merge all this in one 'cause it could produce unexpected behaviour. E.g. while one address is deleting, another one could be created but the addressState would be set to null in success handler

File: packages/dialect-react-sdk/hooks/useAddresses.ts [hunk 8]
```diff
@@ -0,0 +1,398 @@
...
+              dappAddress,
+              enabled: Boolean(dappAddress?.enabled),
+            };
+          })
+        : [],
+    [addresses, dappAddresses]
+  );
+
+  const mergeAddress = useCallback(
+    (data) =>
```
[the remark this answers is not in the document]
Comment (Kiryous) [reply packages/dialect-react-sdk/hooks/useAddresses.ts:145]: yep

File: packages/dialect-react-sdk/hooks/useAddresses.ts (hunk 5 above)
Comment (Kiryous) [reply packages/dialect-react-sdk/hooks/useAddresses.ts:52]: @zhelezkov I added it to support empty object as value — https://stackoverflow.com/a/68208381/10197410, but prob it's not necesary

File: packages/dialect-react-sdk/hooks/useAddresses.ts (hunk 1 above)
Comment (Kiryous) [reply packages/dialect-react-sdk/hooks/useAddresses.ts:334]: nice catch, removed it
